mesonbuild/modules/unstable_wayland.py

```diff
@@ -0,0 +1,121 @@
+# limitations under the License.
+
+import os
+
+from . import ExtensionModule, ModuleReturnValue
+from .. import mlog
+from ..build import CustomTarget
+from ..interpreter.type_checking import NoneType, in_set_validator
+from ..interpreterbase import FeatureNew, typed_pos_args, typed_kwargs, KwargInfo
+from ..mesonlib import File, MesonException, MachineChoice
+
+
+class WaylandModule(ExtensionModule):
+
+    @FeatureNew('wayland module', '0.62.0')
+    def __init__(self, interpreter):
+        super().__init__(interpreter)
+
+        self.protocols_dep = None
+        self.pkgdatadir = None
+        self.scanner_bin = None
+
+        self.methods.update({
+            'scan_xml': self.scan_xml,
+            'find_protocol': self.find_protocol,
+        })
+
+    @typed_pos_args('wayland.scan_xml', varargs=(str, File), min_varargs=1)
+    @typed_kwargs(
+        'wayland.scan_xml',
+        KwargInfo('side', str, default='client', validator=in_set_validator({'client', 'server'})),
+        KwargInfo('scope', str, default='private', validator=in_set_validator({'private', 'public'})),
+    )
+    def scan_xml(self, state, args, kwargs):
+        if self.scanner_bin is None:
+            self.scanner_bin = state.find_program('wayland-scanner', for_machine=MachineChoice.BUILD)
+
+        scope = kwargs['scope']
+        side = kwargs['side']
+
+        xml_files = self.interpreter.source_strings_to_files(args[0])
+        targets = []
+        for xml_file in xml_files:
+            base_name = os.path.basename(xml_file.fname)
+            name, _ = os.path.splitext(base_name)
+
+            code = CustomTarget(
+                f'{name}-protocol',
+                state.subdir,
+                state.subproject,
+                [self.scanner_bin, f'{scope}-code', '@INPUT@', '@OUTPUT@'],
+                [xml_file],
+                [f'{name}-protocol.c'],
+                backend=state.backend,
+            )
+            targets.append(code)
+
+            header = CustomTarget(
+                f'{name}-{side}-protocol',
+                state.subdir,
+                state.subproject,
+                [self.scanner_bin, f'{side}-header', '@INPUT@', '@OUTPUT@'],
+                [xml_file],
+                [f'{name}-{side}-protocol.h'],
+                backend=state.backend,
+            )
+            targets.append(header)
+
+        return ModuleReturnValue(targets, targets)
+
+    @typed_pos_args('wayland.find_protocol', str)
+    @typed_kwargs(
+        'wayland.find_protocol',
+        KwargInfo('state', str, default='stable', validator=in_set_validator({'stable', 'staging', 'unstable'})),
+        KwargInfo('version', (int, NoneType)),
+    )
+    def find_protocol(self, state, args, kwargs):
+        base_name = args[0]
+        xml_state = kwargs['state']
+        version = kwargs['version']
+
+        if (xml_state == 'staging' or xml_state == 'unstable') and version is None:
```

Comment:
```suggestion
        if xml_state in {'staging', 'unstable'} and version is None:
```

Or

```suggestion
        if xml_state != 'stable' and version is None:
```

Either one currently gives technically correct results, the semantics depend on whether you'd expect potential future types to require a version or not. I'd actually expect the latter to be correct, but your current check implies the former.